core/storage/topic/gae_models.py

```diff
@@ -21,6 +21,104 @@
 from google.appengine.ext import ndb
 
 
+class TopicSnapshotMetadataModel(base_models.BaseSnapshotMetadataModel):
+    """Storage model for the metadata for a topic snapshot."""
+    pass
+
+
+class TopicSnapshotContentModel(base_models.BaseSnapshotContentModel):
+    """Storage model for the content of a topic snapshot."""
+    pass
+
+
+class TopicModel(base_models.VersionedModel):
+    """Model for storing Topics.
+
+    This class should only be imported by the topic domain file, the
+    topic services file, and the topic model test file.
+    """
+    SNAPSHOT_METADATA_CLASS = TopicSnapshotMetadataModel
+    SNAPSHOT_CONTENT_CLASS = TopicSnapshotContentModel
+    ALLOW_REVERT = False
+
+    # The name of the topic.
+    name = ndb.StringProperty(required=True, indexed=True)
+    # This includes additional metadata like the description of the topic and
+    # some content paragraphs.
+    description = ndb.StringProperty(indexed=False)
+    # This consists of the list of canonical story ids that are part of
+    # this topic.
+    canonical_story_ids = ndb.StringProperty(repeated=True, indexed=False)
+    # This consists of the list of additional (non-canonical) story ids that
+    # are part of this topic.
+    additional_story_ids = ndb.StringProperty(repeated=True, indexed=False)
+    # This consists of the full list of skill ids that are part of this topic.
+    skill_ids = ndb.StringProperty(repeated=True, indexed=False)
+    # The ISO 639-1 code for the language this question is written in.
+    language_code = ndb.StringProperty(required=True, indexed=True)
+
+
+class TopicCommitLogEntryModel(base_models.BaseCommitLogEntryModel):
+    """Log of commits to topics.
+
+    A new instance of this model is created and saved every time a commit to
+    TopicModel occurs.
+
+    The id for this model is of the form
+    'topic-{{TOPIC_ID}}-{{TOPIC_VERSION}}'.
+    """
+    # The id of the topic being edited.
+    topic_id = ndb.StringProperty(indexed=True, required=True)
+
+    @classmethod
+    def get_commit(cls, topic_id, version):
```

Comment:
It may not be a bad idea to put get_commit() in the superclass (since that behaviour is common to all of these subclasses), and have a method called get_instance_id() that raises NotImplementedError in the superclass and that is overwritten in subclasses. That makes the subclasses contain only customizable functionality, which is cleaner.

Reply:
Done. Moved get_commit into BaseCommitLogEntryModel, and added _get_instance_id() to all the inherited classes to get the unique log id.